File: src/main/java/io/etcd/recipes/examples/basics/SetValueWithTxn.java
/*
 * Copyright © 2019 Paul Ambrose (pambrose@mac.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.etcd.recipes.examples.basics;

import com.google.common.collect.Lists;
import io.etcd.jetcd.Client;
import io.etcd.jetcd.KV;

import java.util.List;

import static io.etcd.recipes.common.ClientUtils.connectToEtcd;
import static io.etcd.recipes.common.KVUtils.delete;
import static io.etcd.recipes.common.KVUtils.getValue;
import static io.etcd.recipes.common.KVUtils.isKeyPresent;
import static io.etcd.recipes.common.KVUtils.putValue;
import static io.etcd.recipes.common.TxnUtils.getDoesExist;
import static io.etcd.recipes.common.TxnUtils.setTo;
import static io.etcd.recipes.common.TxnUtils.transaction;
import static java.lang.String.format;

public class SetValueWithTxn {
    private static final List<String> urls = Lists.newArrayList("http://localhost:2379");
    private static final String path = "/txnexample";
    private static final String keyval = "foobar";

    public static void main(String[] args) {
        try (Client client = connectToEtcd(urls);
             KV kvClient = client.getKVClient()) {

            System.out.println("Deleting keys");
            delete(kvClient, path, keyval);

            System.out.println(format("Key present: %s", isKeyPresent(kvClient, keyval)));
            checkForKey(kvClient);
            System.out.println(format("Key present: %s", isKeyPresent(kvClient, keyval)));
            putValue(kvClient, path, "Something");
            checkForKey(kvClient);
        }
    }

    private static void checkForKey(KV kvClient) {
        transaction(kvClient, (txn) -> {
            txn.If(getDoesExist(path));
            txn.Then(setTo(keyval, format("Key %s found", path)));
            txn.Else(setTo(keyval, format("Key %s not found", path)));
            return txn;
        });

        System.out.println(format("Debug value: %s", getValue(kvClient, keyval, "not_used")));
    }
}

File: src/main/kotlin/io/etcd/recipes/examples/basics/SetValueWithTxn.kt
/*
 * Copyright © 2019 Paul Ambrose (pambrose@mac.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("UndocumentedPublicClass", "UndocumentedPublicFunction")

package io.etcd.recipes.examples.basics

import io.etcd.jetcd.KV
import io.etcd.recipes.common.delete
import io.etcd.recipes.common.doesExist
import io.etcd.recipes.common.etcdExec
import io.etcd.recipes.common.getValue
import io.etcd.recipes.common.isKeyPresent
import io.etcd.recipes.common.putValue
import io.etcd.recipes.common.setTo
import io.etcd.recipes.common.transaction

fun main() {
    val urls = listOf("http://localhost:2379")
    val path = "/txnexample"
    val keyval = "debug"

    fun checkForKey(kvClient: KV) {
        kvClient.transaction {
            If(path.doesExist)
            Then(keyval setTo "Key $path found")
            Else(keyval setTo "Key $path not found")
        }

        println("Debug value: ${kvClient.getValue(keyval, "not_used")}")
    }

    etcdExec(urls) { _, kvClient ->
        println("Deleting keys")
        kvClient.delete(path, keyval)

        println("Key present: ${kvClient.isKeyPresent(keyval)}")
        checkForKey(kvClient)
        println("Key present: ${kvClient.isKeyPresent(keyval)}")
        kvClient.putValue(path, "Something")
        checkForKey(kvClient)
    }
}